src/softmax.jl

```diff
@@ -98,19 +100,19 @@ function logsoftmax!(out::AbstractVecOrMat, xs::AbstractVecOrMat)
             end
             s = zero(eltype(out))
             for i = 1:size(out, 1)
-                s += exp(xs[i, j] - xi_max)
+                s += SLEEF.exp(xs[i, j] - xi_max)
```

Comment:
On Linux with a recent glibc, `libmvec` provides a fast `exp` implementation. LoopVectorization with `@avx` will use this; it will often be faster than the autovectorizer working on the scalar `SLEEFpirates.exp`.
It's worth benchmarking, and of course always worth considering the compile time trade-off as well.

Reply:
Hmm:
```julia
julia> using SLEEFPirates, LoopVectorization
[ Info: Precompiling SLEEFPirates [476501e8-09a2-5ece-8869-fb82de89a1fa]
[ Info: Precompiling LoopVectorization [bdcacae8-1622-11e9-2a5c-532679323890]

julia> function sumexp(x)
           s = zero(eltype(x))
           @inbounds @simd for i ∈ eachindex(x)
               s += SLEEFPirates.exp(x[i])
           end
           s
       end
sumexp (generic function with 1 method)

julia> function sumexpavx(x)
           s = zero(eltype(x))
           @avx for i ∈ eachindex(x)
               s += SLEEFPirates.exp(x[i])
           end
           s
       end
sumexpavx (generic function with 1 method)

julia> x64 = rand(1024);

julia> @btime sumexp($x64)
  4.613 μs (0 allocations: 0 bytes)
1737.405484403936

julia> @btime sumexpavx($x64)
  596.193 ns (0 allocations: 0 bytes)
1737.4054844039383

julia> @btime sum(exp, $x64)
  7.412 μs (0 allocations: 0 bytes)
1737.405484403936

julia> x32 = rand(Float32, 1024);

julia> @btime sumexp($x32)
  245.346 ns (0 allocations: 0 bytes)
1765.1919f0

julia> @btime sumexpavx($x32)
  295.561 ns (0 allocations: 0 bytes)
1765.1919f0

julia> @btime sum(exp, $x32)
  4.657 μs (0 allocations: 0 bytes)
1765.192f0
```
So `SLEEFPirates.exp` in a loop is faster with `Float32`. I'll look into this.

Reply:
With `Float64`, autovectorization fails on Julia 1.4.1.
With `Float32`, avoiding ccall and instead inlining and everything, hoisting the constant loads, and unrolling is faster for longer input vectors.

Reply:
Anyway, worth pointing out that reducing a floating point reduction isn't legal without some sort of annotation (like `@simd`).

Reply:
Thanks! I will change that to `@avx`.